source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -94,10 +97,13 @@ Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& he
   // Either a cache miss or a cache entry that is no longer valid.
   // Check if the new response can be cached.
   if (request_allows_inserts_ && CacheabilityUtils::isCacheableResponse(headers)) {
-    // TODO(#12140): Add date internal header or metadata to cached responses.
     ENVOY_STREAM_LOG(debug, "CacheFilter::encodeHeaders inserting headers", *encoder_callbacks_);
     insert_ = cache_.makeInsertContext(std::move(lookup_));
+    // Add the response time internal header to be copied into the cache
+    headers.setInline(response_time_handle.handle(), date_formatter_.now(time_source_));
     insert_->insertHeaders(headers, end_stream);
+    // Then remove the internal header from the response served to the client
+    headers.removeInline(response_time_handle.handle());
```

Comment:
I don't like the way the response time header is added before inserts/updates to add it to the cached response, then removed before serving to the user. However I can't think of any cleaner (yet efficient) way to do it.

One way to make it cleaner is to copy the response_headers into a new `cached_headers` variable, add the response_time to it, then pass it to the cache. However, copying the response headers is not efficient and does not scale well. Also, copying a header map is not that clean anyway(not a simple assignment). 

Any suggestions here would be appreciated!

Reply:
Why do we need a proto rather than a normal struct that the plug-in decides what to do with it?

Reply:
Because I've worked at Google long enough to think everything should be a proto. :-P
For implementations that store responses remotely, a proto would have the advantage of being easily serialized, with new additions handled seamlessly. For in-process caches, a struct would be slightly more efficient. IMO, the balance leans towards proto. That said, I'll support your choice.